### ISSUE TITLE
`spack install` does not install dependencies as in `spack spec`

### ISSUE THREAD

**Issue by VaiaPatta1985** [LLNL/spack#45578]
### Steps to reproduce

There are differences between `spack spec` and `spack install`. Observed with the following specs:

`spack spec -lI protobuf os=centos7 %gcc@13`
`spack install protobuf os=centos7 %gcc@13`

Details of the result are in the "Error message" section, which contains the output.

### Error message

Command `spack spec -lI protobuf os=centos7 %gcc@13` gave me the following output:

```
Input spec
--------------------------------
 -   protobuf%gcc@13 arch=None-centos7-None

Concretized
--------------------------------
 -   zxeaksu  protobuf@3.25.3%gcc@13.2.0~ipo+shared build_system=cmake build_type=Release generator=make arch=linux-centos7-x86_64
 -   qimn3sq      ^abseil-cpp@20240116.2%gcc@13.2.0~ipo+shared build_system=cmake build_type=Release cxxstd=14 generator=make arch=linux-centos7-x86_64
[+]  gkhtye7      ^cmake@3.27.9%gcc@13.2.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-centos7-x86_64
[+]  uodpt3m          ^curl@8.7.1%gcc@13.2.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-centos7-x86_64
[+]  wyzrf43              ^nghttp2@1.57.0%gcc@13.2.0 build_system=autotools arch=linux-centos7-x86_64
[+]  pzwdno7                  ^diffutils@3.10%gcc@13.2.0 build_system=autotools arch=linux-centos7-x86_64
[+]  nauyh2d              ^openssl@3.3.0%gcc@13.2.0~docs+shared build_system=generic certs=mozilla arch=linux-centos7-x86_64
[+]  x5f4fgs                  ^ca-certificates-mozilla@2023-05-30%gcc@13.2.0 build_system=generic arch=linux-centos7-x86_64
[+]  ymvsmvb              ^perl@5.38.0%gcc@13.2.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-centos7-x86_64
[+]  ygmjnqp                  ^berkeley-db@18.1.40%gcc@13.2.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-centos7-x86_64
[+]  aalli7p                  ^bzip2@1.0.8%gcc@13.2.0~debug~pic+shared build_system=generic arch=linux-centos7-x86_64
[+]  dmx7unp                  ^gdbm@1.23%gcc@13.2.0 build_system=autotools arch=linux-centos7-x86_64
[+]  2zrwzdw                      ^readline@8.2%gcc@13.2.0 build_system=autotools patches=bbf97f1 arch=linux-centos7-x86_64
[+]  dzpsfnx              ^pkgconf@2.2.0%gcc@13.2.0 build_system=autotools arch=linux-centos7-x86_64
[+]  dp4kc7i          ^ncurses@6.5%gcc@13.2.0~symlinks+termlib abi=none build_system=autotools patches=7a351bc arch=linux-centos7-x86_64
[+]  p3ytakl      ^gcc-runtime@13.2.0%gcc@13.2.0 build_system=generic arch=linux-centos7-x86_64
[e]  bvg6gnp      ^glibc@2.34%gcc@13.2.0 build_system=autotools arch=linux-centos7-x86_64
[+]  4qcmieo      ^gmake@4.4.1%gcc@13.2.0~guile build_system=generic arch=linux-centos7-x86_64
[+]  5cp3e7t      ^zlib-ng@2.1.6%gcc@13.2.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=linux-centos7-x86_64
```

The next command was `spack install protobuf os=centos7 %gcc@13`, which installed the following:

```
[+]  5u6jcjh  protobuf@3.25.3%gcc@13.2.0~ipo+shared build_system=cmake build_type=Release generator=make arch=linux-centos7-x86_64
[+]  2ouvbut      ^abseil-cpp@20240116.2%gcc@13.2.0~ipo+shared build_system=cmake build_type=Release cxxstd=14 generator=make arch=linux-centos7-x86_64
[+]  gkhtye7      ^cmake@3.27.9%gcc@13.2.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-centos7-x86_64
[+]  uodpt3m          ^curl@8.7.1%gcc@13.2.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-centos7-x86_64
[+]  wyzrf43              ^nghttp2@1.57.0%gcc@13.2.0 build_system=autotools arch=linux-centos7-x86_64
[+]  pzwdno7                  ^diffutils@3.10%gcc@13.2.0 build_system=autotools arch=linux-centos7-x86_64
[+]  nauyh2d              ^openssl@3.3.0%gcc@13.2.0~docs+shared build_system=generic certs=mozilla arch=linux-centos7-x86_64
[+]  x5f4fgs                  ^ca-certificates-mozilla@2023-05-30%gcc@13.2.0 build_system=generic arch=linux-centos7-x86_64
[+]  ymvsmvb              ^perl@5.38.0%gcc@13.2.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-centos7-x86_64
[+]  ygmjnqp                  ^berkeley-db@18.1.40%gcc@13.2.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-centos7-x86_64
[+]  aalli7p                  ^bzip2@1.0.8%gcc@13.2.0~debug~pic+shared build_system=generic arch=linux-centos7-x86_64
[+]  dmx7unp                  ^gdbm@1.23%gcc@13.2.0 build_system=autotools arch=linux-centos7-x86_64
[+]  2zrwzdw                      ^readline@8.2%gcc@13.2.0 build_system=autotools patches=bbf97f1 arch=linux-centos7-x86_64
[+]  dzpsfnx              ^pkgconf@2.2.0%gcc@13.2.0 build_system=autotools arch=linux-centos7-x86_64
[+]  p3ytakl          ^gcc-runtime@13.2.0%gcc@13.2.0 build_system=generic arch=linux-centos7-x86_64
[+]  dp4kc7i          ^ncurses@6.5%gcc@13.2.0~symlinks+termlib abi=none build_system=autotools patches=7a351bc arch=linux-centos7-x86_64
[+]  sej6ats      ^gcc-runtime@13.2.0%gcc@13.2.0 build_system=generic arch=linux-centos7-x86_64
[e]  dtaqd4e          ^glibc@2.34%gcc@11.4.1 build_system=autotools arch=linux-rocky9-x86_64
[e]  bvg6gnp      ^glibc@2.34%gcc@13.2.0 build_system=autotools arch=linux-centos7-x86_64
[+]  4qcmieo      ^gmake@4.4.1%gcc@13.2.0~guile build_system=generic arch=linux-centos7-x86_64
[+]  5cp3e7t      ^zlib-ng@2.1.6%gcc@13.2.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=linux-centos7-x86_64
```

In other words, a different hash was installed. Notice that `spack install` arbitrarily decided to use a different hash for `abseil-cpp` to the one in the `spack spec` output. Crucially, this led to the use of an unwanted dependency further down (different hash for `gcc-runtime`).

### Information on your system

* **Spack:** 0.22.1
* **Python:** 3.11.7
* **Platform:** linux-rocky9-zen2
* **Concretizer:** clingo
* **Configuration files:** [spack.zip](https://github.com/user-attachments/files/16493987/spack.zip)

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by alalazo** [comment]
I can't reproduce this issue on a simpler setup, I'll try later on a rocky9 machine, using your configuration and will report here.

**Comment by alalazo** [comment]
Hmm, looking at the specs you posted, it seems the issue is that a `glibc` node is labeled `os=rocky9` in the install, whereas in the `spack spec` command all the nodes are `centos7`. Did you install anything in between the two commands? Also, did you try using the `--fresh` option to both `spack spec` and `spack install`?

**Comment by VaiaPatta1985** [comment]
I didn't install anything inbetween; the commands were consecutive, and only 2 minutes apart.

I didn't try with `--fresh`, but whatever the default was, don't `spec` and `install` use the same default? I mean, when I don't give them `--fresh` or `--reuse`, shouldn't they both use the same setting? Since the same parameters were given to `spec` and `install`, I was expecting `install` to do what `spec` showed me.

**Comment by alalazo** [comment]
> I didn't try with --fresh, but whatever the default was, don't spec and install use the same default? I mean, when I don't give them --fresh or --reuse, shouldn't they both use the same setting? [...]

Yes, they should produce the same result, if nothing is installed in-between. I was just wondering if some details were missing from the issue.
